persistent/picklecache.py

```diff
@@ -229,36 +307,104 @@ def debug_info(self):
+        # To avoid mutating datastructures in place or making a copy,
+        # and to work efficiently with both the CFFI ring and the
+        # deque-based ring, we collect the objects and their indexes
+        # up front and then hand them off for ejection.
+        # We don't use enumerate because that's slow under PyPy
+        i = -1
+        to_eject = []
+        for value in self.ring:
+            if self.non_ghost_count <= target and (self.total_estimated_size <= target_size_bytes or not target_size_bytes):
+                break
+            i += 1
+            if value._p_state == UPTODATE:
+                # The C implementation will only evict things that are specifically
+                # in the up-to-date state
+                self._persistent_deactivate_ran = False
+
+                # sweeping an object out of the cache should also
+                # ghost it---that's what C does. This winds up
+                # calling `update_object_size_estimation`.
+                # Also in C, if this was the last reference to the object,
+                # it removes itself from the `data` dictionary.
+                # If we're under PyPy or Jython, we need to run a GC collection
+                # to make this happen...this is only noticeable though, when
+                # we eject objects. Also, note that we can only take any of these
+                # actions if our _p_deactivate ran, in case of buggy subclasses.
+                # see _persistent_deactivate_ran
+
+                value._p_deactivate()
+                if (self._persistent_deactivate_ran
+                    # Test-cases sneak in non-Persistent objects, sigh, so naturally
+                    # they don't cooperate (without this check a bunch of test_picklecache
+                    # breaks)
+                    or not isinstance(value, _SWEEPABLE_TYPES)):
+                    to_eject.append((i, value))
+                    self.non_ghost_count -= 1
+
+        ejected = len(to_eject)
+        if ejected:
+            self.ring.delete_all(to_eject)
+        del to_eject # Got to clear our local if we want the GC to get the weak refs
+
+        if ejected and _SWEEP_NEEDS_GC:
+            # See comments on _SWEEP_NEEDS_GC
+            gc.collect()
+        return ejected
+
+    @_sweeping_ring
     def _invalidate(self, oid):
         value = self.data.get(oid)
         if value is not None and value._p_state != GHOST:
             value._p_invalidate()
-            node = self.ring.next
-            while True:
-                if node is self.ring:
-                    break  # pragma: no cover belt-and-suspenders
-                if node.object is value:
-                    node.prev.next, node.next.prev = node.next, node.prev
-                    break
-                node = node.next
+            was_in_ring = self.ring.delete(value)
+            assert was_in_ring, "Ring corruption: invalidated a non-ghost not in ring"
```

Comment:
Why an `assert` here -- do you really want the check compiled away when optimized?